dc/s2n-quic-dc/src/socket/recv/descriptor.rs

```diff
@@ -0,0 +1,495 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::msg::{addr::Addr, cmsg};
+use core::fmt;
+use s2n_quic_core::inet::ExplicitCongestionNotification;
+use std::{
+    io::IoSliceMut,
+    marker::PhantomData,
+    ptr::NonNull,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Weak,
+    },
+};
+use tracing::trace;
+
+/// Callback which releases a descriptor back into the free list
+pub(super) trait FreeList: 'static + Send + Sync {
+    fn free(&self, descriptor: Descriptor);
+}
+
+/// A handle to various parts for the descriptor group instance
+pub(super) struct Memory {
+    capacity: u16,
+    references: AtomicUsize,
+    free_list: Weak<dyn FreeList>,
+    #[allow(dead_code)]
+    region: Box<dyn 'static + Send + Sync>,
```

Comment:
This is just `Region`, right? Why are we putting it in a Box?

Reply:
This is to support both the `pool` implementation as well as a future `AF_XDP` UMEM allocation. So i just needed something that is droppable once `Memory` is freed.

Reply:
This is to support both the `pool` implementation as well as a future `AF_XDP` UMEM allocation. So i just needed something that is droppable once `Memory` is freed.